### PR TITLE
Security patch for Autogluon 0.3.2 and bug fix for test cuda path failures

### PR DESCRIPTION
**Pull request by kenny-ezirim**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: data/ignore_ids_safety_scan.json [hunk 1]
```diff
@@ -692,7 +694,9 @@
                 "44716": "[DISPUTED] The vendor does not agree this is a vulnerability; the negative dimensions can only be created by an already privileged user (or internally).",
                 "44715": "[DISPUTED] While correct that validation is missing, an error can only occur due to an exhaustion of memory. If the user can exhaust memory, they are already privileged. Further, it should be practically impossible to construct an attack which can target the memory exhaustion to occur at exactly this place.",
                 "43581": "pytorch-lightning stable release (1.6.0) is not available",
-                "43752": "pytorch-lightning stable release (1.6.0) is not available"
+                "43752": "pytorch-lightning stable release (1.6.0) is not available",
+                "48597": "This is a false-positive. The pillow dependency is patched inside the container.",
```
Comment (jeet4320) [inline data/ignore_ids_safety_scan.json:698]: can you add which pillow dependency

File: test/test_utils/__init__.py [hunk 2]
```diff
@@ -1510,6 +1510,35 @@ def get_python_version_from_image_uri(image_uri):
     python_version = python_version_search.group()
     return "py36" if python_version == "py3" else python_version
 
+def construct_buildspec_path(dlc_path, framework_path, buildspec, framework_version):
```
Comment (jeet4320) [inline test/test_utils/__init__.py:1513]: any reason to add this

Comment (kenny-ezirim) [reply test/test_utils/__init__.py:1513]: Address an issue with deriving framework short version. Previous implementation introduced a bug with the framework_short_version having a NoneType for AG.

File: data/ignore_ids_safety_scan.json (hunk 1 above)
Comment (kenny-ezirim) [reply data/ignore_ids_safety_scan.json]: Added

File: test/test_utils/__init__.py [hunk 3]
```diff
@@ -1510,6 +1510,35 @@ def get_python_version_from_image_uri(image_uri):
...
+        if matched:
+            version = ""
+            versions = []
+            for match in matched.groups():
+                if match:
+                    version = f'{version}{match.replace(".","-")}'
+                    versions.append(version)
+
+            # skip search with only major version number
+            for v in reversed(versions):
```
Comment (arjkesh) [inline test/test_utils/__init__.py:1536]: what does "versions" look like? for 1.2.3, are you expecting it to look like [1-2-3, 1-2, 1]? Can you confirm it doesn't look like [-3, -2, 1]

File: test/test_utils/__init__.py [hunk 4]
```diff
@@ -1510,6 +1510,35 @@ def get_python_version_from_image_uri(image_uri):
...
+    a specific version file for the framework being tested. Possible options include:
+    [buildspec-[Major]-[Minor]-[Patch].yml, buildspec-[Major]-[Minor].yml, buildspec-[Major].yml, buildspec.yml]
+    :param dlc_path: path to the DLC test folder
+    :param framework_path: Framework folder name
+    :param buildspec: buildspec file name
+    :param framework_version: default (long) framework version name
+    """
+    if framework_version:
+        # pattern matches for example 0.3.2 or 22.3
+        pattern = r"^(\d+)?(\.\d+)?(\*|\.\d+)$"
```
[the remark this answers is not in the document]
Comment (kenny-ezirim) [reply test/test_utils/__init__.py:1525]: We can do without the *. Does not apply to our case.

File: test/test_utils/__init__.py [hunk 5]
```diff
@@ -1510,6 +1510,35 @@ def get_python_version_from_image_uri(image_uri):
...
+    """
+    Construct a relative path to the buildspec yaml file by iterative checking on the existence of
+    a specific version file for the framework being tested. Possible options include:
+    [buildspec-[Major]-[Minor]-[Patch].yml, buildspec-[Major]-[Minor].yml, buildspec-[Major].yml, buildspec.yml]
+    :param dlc_path: path to the DLC test folder
+    :param framework_path: Framework folder name
+    :param buildspec: buildspec file name
+    :param framework_version: default (long) framework version name
+    """
+    if framework_version:
```
[the remark this answers is not in the document]
Comment (kenny-ezirim) [reply test/test_utils/__init__.py:1523]: Yes. The default path can be built without using the framework version. See last line of function.

File: test/test_utils/__init__.py [hunk 6]
```diff
@@ -1510,6 +1510,35 @@ def get_python_version_from_image_uri(image_uri):
...
+        # pattern matches for example 0.3.2 or 22.3
+        pattern = r"^(\d+)?(\.\d+)?(\*|\.\d+)$"
+        matched = re.search(pattern, framework_version)
+        if matched:
+            version = ""
```
[the remark this answers is not in the document]
Comment (kenny-ezirim) [reply test/test_utils/__init__.py:1528]: We need to concatenate the version string from right to left. So given [1, .0, .2], version will be in this order: [1,1.0,1.0.2]
if we do version = f'{match.replace(...)}', we will get [1,-0,-2], which we don't want.

File: test/test_utils/__init__.py (hunk 3 above)
Comment (kenny-ezirim) [reply test/test_utils/__init__.py:1536]: 1.2.3 will go as follows:
"1.2.3" ---regex search--> [1,.2,.3] ---transformed----> [1,1-2,1-2-3] ----reversed----> [1-2-3,1-2,1].  This is confirmed.

File: autogluon/buildspec.yml [hunk 7]
```diff
@@ -1,8 +1,8 @@
 account_id: &ACCOUNT_ID <set-$ACCOUNT_ID-in-environment>
 region: &REGION <set-$REGION-in-environment>
```
[the remark this answers is not in the document]
Comment (kenny-ezirim) [reply autogluon/buildspec.yml]: Reverted.

File: test/test_utils/__init__.py [hunk 8]
```diff
@@ -1510,6 +1510,35 @@ def get_python_version_from_image_uri(image_uri):
...
+            versions = []
+            for match in matched.groups():
+                if match:
+                    version = f'{version}{match.replace(".","-")}'
+                    versions.append(version)
+
+            # skip search with only major version number
```
Comment (arjkesh) [inline test/test_utils/__init__.py:1535]: Nit: Given the comment in the resolved conversation below, how is the only-major-version number skipped? Is this still a relevant comment?

File: test/test_utils/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kenny-ezirim) [reply test/test_utils/__init__.py:1523]: Framework version is derived from image uri. In the case where framework version cannot be retrieved from image, the framework version can be empty. This is very unlikely given the naming convention used for DLC images. See function get_framework_and_version_from_tag for how to get version of an image uri. I don't think we need to provide a default value since we expect that framework version will always be provided.

File: test/test_utils/__init__.py (hunk 8 above)
Comment (kenny-ezirim) [reply test/test_utils/__init__.py:1535]: The major version is not skipped. We expect the framework version to be composed of major, minor and patch. According to @saimidu, we start with a narrower version number to a more broader version number. For instance, in the case of 1.2.3, we should search for the buildspec file in the following order: 
 buildspec-1-2-3.yml -> buildspec-1-2.yml -> buildspec-1.yml -> buildspec.yml (default buildspec file).

File: test/test_utils/__init__.py [hunk 9]
```diff
@@ -1510,6 +1510,35 @@ def get_python_version_from_image_uri(image_uri):
...
+    :param buildspec: buildspec file name
+    :param framework_version: default (long) framework version name
+    """
+    if framework_version:
+        # pattern matches for example 0.3.2 or 22.3
+        pattern = r"^(\d+)?(\.\d+)?(\.\d+)$"
+        matched = re.search(pattern, framework_version)
+        if matched:
+            version = ""
+            versions = []
```
Comment (saimidu) [inline test/test_utils/__init__.py:1529]: nit: I'd recommend renaming either `version` or `versions`, because the variable names can get confusing for a new reader on this code-base.

Perhaps you can change `version` to `constructed_version` or `buildspec_version`, or you could change `versions` to `versions_to_search`.

File: test/test_utils/__init__.py (hunk 8 above)
Comment (kenny-ezirim) [reply test/test_utils/__init__.py:1535]: The comment is misplaced. I am removing because it no longer applies in the current implementation.

File: test/test_utils/__init__.py [hunk 10]
```diff
@@ -1510,6 +1510,35 @@ def get_python_version_from_image_uri(image_uri):
...
+    a specific version file for the framework being tested. Possible options include:
+    [buildspec-[Major]-[Minor]-[Patch].yml, buildspec-[Major]-[Minor].yml, buildspec-[Major].yml, buildspec.yml]
+    :param dlc_path: path to the DLC test folder
+    :param framework_path: Framework folder name
+    :param buildspec: buildspec file name
+    :param framework_version: default (long) framework version name
+    """
+    if framework_version:
+        # pattern matches for example 0.3.2 or 22.3
+        pattern = r"^(\d+)?(\.\d+)?(\.\d+)$"
```
[the remark this answers is not in the document]
Comment (kenny-ezirim) [reply test/test_utils/__init__.py:1525]: The suggested change is correct. The ? mark should be after the second and third digits. I check the implication of changing the order.

File: test/dlc_tests/sanity/test_pre_release.py [hunk 11]
```diff
@@ -670,7 +671,9 @@ def test_cuda_paths(gpu):
 
     # Get cuda, framework version, python version through regex
     cuda_version = re.search(r"-(cu\d+)-", image).group(1)
-    framework_short_version = None
+    
+    framework_short_version = re.match(r"(\d+.\d+)", framework_version).group(1)
```
[the remark this answers is not in the document]
Comment (kenny-ezirim) [reply test/dlc_tests/sanity/test_pre_release.py:675]: Corrected.

File: test/test_utils/__init__.py (hunk 9 above)
Comment (kenny-ezirim) [reply test/test_utils/__init__.py:1529]: Recommendation adopted.

File: test/test_utils/__init__.py [hunk 12]
```diff
@@ -1510,6 +1510,35 @@ def get_python_version_from_image_uri(image_uri):
...
+        matched = re.search(pattern, framework_version)
+        if matched:
+            version = ""
+            versions = []
+            for match in matched.groups():
+                if match:
```
[the remark this answers is not in the document]
Comment (kenny-ezirim) [reply test/test_utils/__init__.py:1531]: It is possible to get None for a group that is not matched, especially if the patch or minor versions are not provided in the version number. For instance, 1.2 will produce 1,.2,None and 1 will produce 1,None, None.

File: test/test_utils/__init__.py [hunk 13]
```diff
@@ -1510,6 +1510,35 @@ def get_python_version_from_image_uri(image_uri):
...
+    :param dlc_path: path to the DLC test folder
+    :param framework_path: Framework folder name
+    :param buildspec: buildspec file name
+    :param framework_version: default (long) framework version name
+    """
+    if framework_version:
+        # pattern matches for example 0.3.2 or 22.3
+        pattern = r"^(\d+)?(\.\d+)?(\.\d+)$"
+        matched = re.search(pattern, framework_version)
+        if matched:
```
[the remark this answers is not in the document]
Comment (kenny-ezirim) [reply test/test_utils/__init__.py:1527]: This is correct. Also, we should check if the buildspec.yml exists before defaulting to it. We might have a case where the buildspec file does not exist at all.

**Review by saimidu — APPROVED** [review]

File: test/test_utils/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (arjkesh) [reply test/test_utils/__init__.py:1523]: Sure, we can look at this as addressing an edge case
